lib/spack/spack/test/detection.py

```diff
@@ -0,0 +1,208 @@
+    cmake_cfg = pkgs_cfg['cmake']
+    cmake_externals = cmake_cfg['externals']
+
+    assert {'spec': 'cmake@1.foo', 'prefix': '/x/y1/'} in cmake_externals
+    assert {'spec': 'cmake@3.17.2', 'prefix': '/x/y2/'} in cmake_externals
+
+
+def test_get_executables(working_env, mock_executable):
+    cmake_path1 = mock_executable("cmake", output="echo cmake version 1.foo")
+    os.environ['PATH'] = ':'.join([os.path.dirname(cmake_path1)])
+    path_to_exe = spack.detection.executables_in_path()
+    assert path_to_exe[cmake_path1] == 'cmake'
+
+
+def test_find_external_merge(mutable_config, mutable_mock_repo):
+    """Check that 'spack find external' doesn't overwrite an existing spec
+    entry in packages.yaml.
+    """
+    pkgs_cfg_init = {
+        'find-externals1': {
+            'externals': [{
+                'spec': 'find-externals1@1.1',
+                'prefix': '/preexisting-prefix/'
+            }],
+            'buildable': False
+        }
+    }
+
+    mutable_config.update_config('packages', pkgs_cfg_init)
+    entries = [
+        spack.detection.DetectedPackage(
+            spack.spec.Spec.from_detection('find-externals1@1.1'), '/x/y1/'
+        ),
+        spack.detection.DetectedPackage(
+            spack.spec.Spec.from_detection('find-externals1@1.2'), '/x/y2/'
+        )
+    ]
+    pkg_to_entries = {'find-externals1': entries}
+    scope = spack.config.default_modify_scope('packages')
+    spack.detection.update_configuration(pkg_to_entries, scope=scope, buildable=True)
+
+    pkgs_cfg = spack.config.get('packages')
+    pkg_cfg = pkgs_cfg['find-externals1']
+    pkg_externals = pkg_cfg['externals']
+
+    assert {'spec': 'find-externals1@1.1',
+            'prefix': '/preexisting-prefix/'} in pkg_externals
+    assert {'spec': 'find-externals1@1.2',
+            'prefix': '/x/y2/'} in pkg_externals
+
+
+def candidate_packages():
+    """Return the list of packages with a corresponding
+    detection_test.yaml file.
+    """
+    # Directories where we have repositories
+    repo_dirs = [spack.util.path.canonicalize_path(x)
+                 for x in spack.config.get('repos')]
+
+    # Compute which files need to be tested
+    to_be_tested = []
+    for repo_dir in repo_dirs:
+        pattern = os.path.join(
+            repo_dir, 'packages', '**', 'detection_test.yaml'
+        )
+        pkgs_with_tests = [os.path.basename(os.path.dirname(x))
+                           for x in glob.glob(pattern)]
+        to_be_tested.extend(pkgs_with_tests)
+
+    return to_be_tested
+
+
+@pytest.mark.detection
+@pytest.mark.parametrize('package_name', candidate_packages())
+def test_package_detection(mock_executable, package_name):
+    def detection_tests_for(pkg):
+        pkg_dir = os.path.dirname(
+            spack.repo.path.filename_for_package_name(pkg)
+        )
+        detection_data = os.path.join(pkg_dir, 'detection_test.yaml')
+        with open(detection_data) as f:
+            return spack.util.spack_yaml.load(f)
```

Comment:
This should move outside of this test, there's no reason for it to be nested

Reply:
Done